specs/agents/metrics-health.md

```diff
@@ -0,0 +1,129 @@
+ - `agent.events.requests.bytes`: `COUNTER` of the approximate number of bytes on the wire sent to the Intake API of the APM
+
+All these metrics MUST have the label `success`, which can have the values `true` or `false`. A request is counted with `success=true` iff the APM Server responded with `2xx`.
+
+The metric `agent.events.requests.bytes` does not need to represent the exact network usage.
+Instead the number of compressed bytes within the request body can be used as approximation.
+
+All event request metrics can be disabled via the `disable_metrics` configuration.
+
+### Agent Overhead Metrics
+
+If possible with a reasonable amount of runtime and implementation overhead, agents SHOULD expose the following metrics for approximating their own resource usage:
+
+ - `agent.background.cpu.overhead.pct`: The fraction of the process cpu usage caused by the agent in a range [0-1] within the last reporting interval.
+ - `agent.background.cpu.total.pct`: The approximate CPU usage caused by the agent. Derived by multiplying `agent.background.cpu.overhead.pct` with `system.process.cpu.total.norm.pct`.
+ - `agent.background.memory.allocation.bytes`: `COUNTER` of the number of bytes allocated in the main memory by the agent.
+ - `agent.background.threads.count`: Number of currently running agent background threads at the time of reporting.
+
+The `agent.background.threads.count` only applies to agents in whose language the concept of `threads` is present and used by the agent for background tasks.
+
+Agents SHOULD add the label `task` to all these metrics. The values for `task` can be freely chosen by each agent implementation. The label values should be chosen in a way which allows the resource usages to be assigned to specific parts of the agent implementation (e.g. thread names as `task`s).
+
+In most cases, it will not be feasible to capture the overhead of every action the agent does due to the overhead of the measurement itself. E.g. for the Java agent it is not feasible to measure the overhead of the code added to user application methods via instrumentation. Instead, only the overhead of agent-owned threads is measured.
+
+The `agent.background.cpu.overhead.pct` can be computed by measuring the increase of thread CPU time of a given task ([this method](https://docs.oracle.com/javase/7/docs/api/java/lang/management/ThreadMXBean.html#getCurrentThreadCpuTime()) for Java) and dividing it by the increase of the total process CPU time ([this method](https://docs.oracle.com/javase/7/docs/jre/api/management/extension/com/sun/management/OperatingSystemMXBean.html#getProcessCpuTime()) for Java) within the same period.
+
+All agent overhead metrics are enabled/disabled via the `agent_background_overhead_metrics` configuration:
+
+|                |   |
+|----------------|---|
+| Type           | `boolean` |
+| Default        | `false` |
+| Dynamic        | `false` |
+| Central config | `false` |
```

Comment:
> maybe another idea: have we considered changing the default of disable_metrics by adding agent.background.* to it for agents which implement overhead health metrics?

I considered it, but it reminded me of why the `experimental`-flag was separated from the `disable_instrumentations` configuration in the Java agent: If we add `agent.background.*` as a default to `disable_metrics` it is very easy for users to enable the overhead metrics by accident.
E.g. a user wants to disable some micrometer metrics, so he changes `disable_metrics=my.metrics.*`. Because he did not add the default `agent.background.*` as well, this unexpectedly enables the overhead metrics.

> I can imagine -especially for critical production apps- that users would love to turn on agent health metrics collection for some time (e.g. if they feel something is not ok with the agent), collect the agent health data, and then turn it off. All without restarting the agent (which typically means restarting their app).
>
> But this applies to both the overhead metrics (which are opt-in) and to the queue metrics (which are handled by disable_metrics if I understand it correctly). My point with this: I think if we make it dynamic (which would be useful), then all health metrics should be dynamic (not only one set of them).

I definitly agree with that it would be nice to have these configuration options dynamic, but thought aswell that it is an enhancement we could add later on. It would for example be great and simplify the configuration interface to just make `disable_metrics` dynamic in general.

Reply:
Summary of the discussion in the java-agent-weekly:

We for now go with the additional configuration flag and later on remove / hide it when we have a better way of dealing with defaults of multi-value configs in general.